design/004-IoT-Adoption-Watcher.md

```diff
@@ -0,0 +1,248 @@
+      states.
+
+   2. Main component should be notified by the watcher, when a state is
+      registered or progressed on the blockchain.
+
+4. If the state registered on the blockchain and is not the latest off-chain
+   state known to the watcher, then it must register the latest state on the
+   blockchain.
+
+### Interfaces of the watcher
+
+In order to realize the desired functionalities, watcher should provide the
+following interfaces:
+
+
+```go
+type Watcher struct{}
+
+// For initalizing the watcher.
+func NewWatcher(rs RegisterSubscriber) *Watcher {...}
+
+// For interacting with the blockchain.
+type RegisterSubscriber interface {
+    Subscriber
+    Registerer
+}
+
+// For registering off-chain state for a given ledger channel,
+// along with the off-chain states of its sub-channels and virtual channels.
+type Registerer interface {
+    Register(State, SubStates)
+}
+
+// For receiving registered and progressed events for a given channel ID.
+type Subscriber interface {
+    Subscribe(ChannelID)
+}
+
+// For registering/de-registering a channel ID with/from the watcher.
+func (w *Watcher) Register(ChannelID, OffChainStatesSub, OnChainEventsPub) {...}
+func (w *Watcher) Deregister(ChannelID) {...}
+
+// For receiving off-chain states from the main component.
+type OffChainStatesSub interface {
+	Next() OffChainState
+}
+
+// For sending on-chain events to the main component.
+type OnChainEventsPub inteface {
+    // OnChainEvent can be a registered event or a progressed event.
+    Publish(OnChainEvent)
+}
+
+// Interfaces for the main component to interact with the watcher.
+
+// For receiving on-chain events from the watcher.
+type OnChainEventsSub interface {
+	Next() OnChainEvent
+}
+
+// For sending off-chain states to the watcher.
+type OffChainStatesPub inteface {
+	Publish(OffChainState)
+}
+
+// For shutting down the watcher
+func (w *Watcher) Shutdown()
+```
+
+The pub-sub instances for off-chain states and on-chain events are initialized
+by the main component and passed on to the watcher. Because,
+
+1. Initial implementation can be a one channel to one watcher pub-sub.
+
+2. Later, if one channel wants to use multiple watcher instances (because one
+ watcher could not provide expected level of availability), then the one
+ channel to multiple watcher pub-sub can also be implemented.
+
+On shutdown call, the watcher must,
+1. Stop watching for all the channel IDs registered with it.
+2. Complete any on-going refutation process.
+3. Close the `OffChainSub` and `OnChainEventsPub` on each of the channels
+  registered with it. This ensures that the channels are notified of the
+  watcher shutdown.
+4. Finally, it should stop.
+
+### Interaction between the watcher and the main component
+
+Interaction between the watcher and main component based using the above
+interfaces is shown in the diagram below.
+
+![Interaction between the main component and watcher](004/watcher.svg)
```

Comment:
Can you insert the function names defined above? Otherwise, it may not be clear what you refer to.

The first step should be to create the watcher?